File: PravahDridh/frontend/src/api/graph.ts
import { apiClient, unwrapResponse } from './client';
import { StandardResponse } from '../types/common';

export interface GraphNode {
  id: string;
  label: string;
  type: string;
  subType?: string;
  properties: Record<string, any>;
  isSuspicious: boolean;
  riskScore?: number;
  severity?: string;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  type: string;
  label?: string;
  isSuspicious: boolean;
  amount?: number;
  timestamp?: string;
}

export interface KnowledgeGraphResponse {
  nodes: GraphNode[];
  edges: GraphEdge[];
  stats: Record<string, number>;
}

export const graphApi = {
  getGraph: async (limit: number = 200): Promise<KnowledgeGraphResponse> => {
    return unwrapResponse<KnowledgeGraphResponse>(
      apiClient.get<StandardResponse<KnowledgeGraphResponse>>('/graph', { params: { limit } })
    );
  },

  getCaseGraph: async (caseId: string): Promise<KnowledgeGraphResponse> => {
    return unwrapResponse<KnowledgeGraphResponse>(
      apiClient.get<StandardResponse<KnowledgeGraphResponse>>(`/graph/case/${caseId}`)
    );
  },
};

File: PravahDridh/backend/app/schemas/graph.py
import uuid
from typing import List, Optional, Dict, Any
from pydantic import BaseModel


class GraphNode(BaseModel):
    id: str
    label: str
    type: str  # Complaint, Account, Transaction, ATM, Investigation, Alert, Prediction
    subType: Optional[str] = None
    properties: Dict[str, Any] = {}
    isSuspicious: bool = False
    riskScore: Optional[float] = None
    severity: Optional[str] = None


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    type: str  # HAS_TRANSACTION, FROM_ACCOUNT, TO_ACCOUNT, USES, OCCURRED_AT, CONNECTED_TO, GENERATED_FOR, TARGETS, INVESTIGATES
    label: Optional[str] = None
    isSuspicious: bool = False
    amount: Optional[float] = None
    timestamp: Optional[str] = None


class KnowledgeGraphResponse(BaseModel):
    nodes: List[GraphNode]
    edges: List[GraphEdge]
    stats: Dict[str, int]
